File: Trials/Network/RUDP/segment/DataSegment.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

/*
 *  DATa Segment
 *
 *   0 1 2 3 4 5 6 7 8            15
 *  +-+-+-+-+-+-+-+-+---------------+
 *  |0|1|0|0|0|0|0|0|       6       |
 *  +-+-+-+-+-+-+-+-+---------------+
 *  | Sequence #    |   Ack Number  |
 *  +---------------+---------------+
 *  |           Checksum            |
 *  +---------------+---------------+
 *  | ...                           |
 *  +-------------------------------+
 *
 */
namespace RUDP
{
    class DataSegment : Segment
    {
        public override string GetSegmentType()
        {
            return "DAT";
        }

        public byte[] Data
        {
            get;
            private set;
        }

        public DataSegment()
        {
        }

        public DataSegment(
            int sequenceNumber,
            int acknowledgedNumber,
            byte[] bytes,
            int offset,
            int length)
            : base(Type.ACK, sequenceNumber, RUDP_HEADER_LEN)
        {
            SetAcknowledgedNumber(acknowledgedNumber);

            this.Data = new byte[length];
            System.Array.Copy(bytes, offset, this.Data, 0, length);
        }

        public override void WriteBytes(Stream wr)
        {
            base.WriteBytes(wr);

            // For CheckSum
            wr.WriteByte(0);
            wr.WriteByte(0);

            wr.Write(this.Data, 0, this.Data.Length);
        }

        public override void ReadBytes(byte[] bytes, int offset, int length)
        {
            base.ReadBytes(bytes, offset, length);

            if (!CheckFlag(Type.ACK))
            {
                throw new ArgumentOutOfRangeException();
            }

            var dataLength = bytes.Length - RUDP_HEADER_LEN;
            this.Data = new byte[dataLength];

            System.Array.Copy(
                bytes,
                offset + RUDP_HEADER_LEN,
                this.Data,
                0,
                this.Data.Length);
        }
    }
}

File: Libs/source/network/rudp/segment/DataSegment.cpp
#include "network/rudp/segment/DataSegment.h"

/*
*  DATa Segment
*
*   0 1 2 3 4 5 6 7 8            15
*  +-+-+-+-+-+-+-+-+---------------+
*  |0|1|0|0|0|0|0|0|       6       |
*  +-+-+-+-+-+-+-+-+---------------+
*  | Sequence #    |   Ack Number  |
*  +---------------+---------------+
*  |           Checksum            |
*  +---------------+---------------+
*  | ...                           |
*  +-------------------------------+
*
*/
namespace izanagi {
namespace net {
    Segment* DataSegment::Create(
        IMemoryAllocator* allocator,
        IZ_INT sequenceNumber,
        IZ_INT acknowledgedNumber,
        IZ_UINT8* bytes,
        IZ_UINT offset,
        IZ_UINT length)
    {
        void* p = ALLOC(allocator, sizeof(DataSegment));
        IZ_ASSERT(p);

        auto ret = new(p)DataSegment(
            allocator,
            sequenceNumber,
            acknowledgedNumber,
            bytes,
            offset, length);

        return ret;
    }

    DataSegment::DataSegment(
        IMemoryAllocator* allocator,
        IZ_INT sequenceNumber,
        IZ_INT acknowledgedNumber,
        IZ_UINT8* bytes,
        IZ_UINT offset,
        IZ_UINT length)
        : Segment(Type::ACK, sequenceNumber, RUDP_HEADER_LEN)
    {
        m_allocator = allocator;

        SetAcknowledgedNumber(acknowledgedNumber);

        m_Data = ALLOC(m_allocator, length);
        memcpy(m_Data, bytes + offset, length);

        m_Size = length;
    }

    DataSegment::~DataSegment()
    {
        if (m_Data) {
            FREE(m_allocator, m_Data);
        }
    }

    void DataSegment::WriteBytes(IOutputStream* wr)
    {
        Segment::WriteBytes(wr);

        // For CheckSum
        wr->Write(0);
        wr->Write(0);

        wr->Write(m_Data, 0, m_Size);
    }

    void DataSegment::ReadBytes(
        IMemoryAllocator* allocator,
        IZ_UINT8* bytes, IZ_UINT offset, IZ_UINT length)
    {
        IZ_ASSERT(m_allocator == nullptr || m_allocator == allocator);

        Segment::ReadBytes(allocator, bytes, offset, length);

        IZ_ASSERT(CheckFlag(Type::ACK));

        auto dataLength = length - RUDP_HEADER_LEN;
        m_Data = ALLOC(allocator, dataLength);

        if (m_allocator == nullptr) {
            m_allocator = allocator;
        }

        memcpy(
            m_Data,
            bytes + offset + RUDP_HEADER_LEN,
            dataLength);

        m_Size = dataLength;
    }
}   // namespace net
}   // namespace izanagi
